File: 351004/Zalivako/DistributedComputing/src/Infrastructure/Kafka/KafkaConsumer.cs
using Confluent.Kafka;
using Core.Entities;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Configuration;
using System.Text.Json;
using Application.Interfaces;
using Microsoft.Extensions.Caching.Distributed;

namespace Infrastructure.Kafka
{
    public class KafkaPostConsumer : BackgroundService
    {
        private readonly IConsumer<string, string> _consumer;
        private readonly IServiceScopeFactory _serviceScopeFactory;
        private readonly ILogger<KafkaPostConsumer> _logger;
        private readonly string _outTopic;

        public KafkaPostConsumer(IConfiguration configuration,
                                 IServiceScopeFactory serviceScopeFactory,
                                 ILogger<KafkaPostConsumer> logger)
        {
            _serviceScopeFactory = serviceScopeFactory;
            _logger = logger;

            var config = new ConsumerConfig
            {
                BootstrapServers = configuration["Kafka:BootstrapServers"],
                GroupId = "publisher-post-status-group",
                AutoOffsetReset = AutoOffsetReset.Earliest,
                EnableAutoCommit = true
            };
            _consumer = new ConsumerBuilder<string, string>(config).Build();
            _outTopic = configuration["Kafka:OutTopic"] ?? "OutTopic";
        }

        protected override Task ExecuteAsync(CancellationToken stoppingToken)
        {
            _consumer.Subscribe(_outTopic);
            return Task.Run(async () =>
            {
                while (!stoppingToken.IsCancellationRequested)
                {
                    try
                    {
                        var consumeResult = _consumer.Consume(stoppingToken);
                        await ProcessMessageAsync(consumeResult.Message.Value, stoppingToken);
                    }
                    catch (OperationCanceledException) { break; }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Error consuming Kafka message");
                    }
                }
            }, stoppingToken);
        }

        private async Task ProcessMessageAsync(string messageValue, CancellationToken cancellationToken)
        {
            try
            {
                var statusUpdate = JsonSerializer.Deserialize<PostStatusUpdate>(messageValue);
                if (statusUpdate == null) return;

                using var scope = _serviceScopeFactory.CreateScope();
                var postRepo = scope.ServiceProvider.GetRequiredService<IPostRepository>();
                var cache = scope.ServiceProvider.GetRequiredService<IDistributedCache>();

                var post = await postRepo.GetByIdAsync(statusUpdate.PostId, cancellationToken);
                if (post != null)
                {
                    post.State = Enum.Parse<PostState>(statusUpdate.NewState);
                    await postRepo.UpdateAsync(post, cancellationToken);

                    // Инвалидация кэша для этого поста и общего списка
                    await cache.RemoveAsync("posts_all");
                    await cache.RemoveAsync($"post_{post.Id}");

                    _logger.LogInformation("Updated post {PostId} state to {State}", post.Id, post.State);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing message: {Message}", messageValue);
            }
        }

        public override void Dispose()
        {
            _consumer?.Dispose();
            base.Dispose();
        }

        private class PostStatusUpdate
        {
            public long PostId { get; set; }
            public string NewState { get; set; } = string.Empty;
        }
    }
}

File: 351003/Sologub/src/apps/notes/kafka_consumer.py
import json
import os
import threading
import logging
import time

from kafka import KafkaConsumer

logger = logging.getLogger(__name__)

KAFKA_BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
OUT_TOPIC = "OutTopic"


def start_consumer():
    def _run():
        consumer = None
        while True:
            try:
                consumer = KafkaConsumer(
                    OUT_TOPIC,
                    bootstrap_servers=KAFKA_BOOTSTRAP,
                    group_id="publisher-group",
                    value_deserializer=lambda m: json.loads(m.decode("utf-8")),
                    auto_offset_reset="earliest",
                    enable_auto_commit=True,
                )
                break
            except Exception as e:
                logger.warning(f"Waiting for Kafka: {e}")
                time.sleep(3)

        logger.info("Publisher Kafka consumer started (OutTopic)")
        for message in consumer:
            try:
                data = message.value
                action = data.get("action")
                note_id = data.get("id")
                state = data.get("state", "")
                logger.info(f"OutTopic: action={action} id={note_id} state={state}")
            except Exception as e:
                logger.error(f"Error processing OutTopic message: {e}")

    thread = threading.Thread(target=_run, daemon=True)
    thread.start()
